### PR TITLE
Add drag and drop demo for tree grid

### PR DESCRIPTION
**Pull request by shadikhani**
1- I didn't use the sample code. I created a simple drag and drop example.
2- This demo has a problem that seems to be a bug in Flow. It does not work when dragging the last item of a parent. I will create an issue later.

Fix https://github.com/vaadin/90min-dx-tasks/issues/103

### CODE REVIEW

File: vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java [hunk 1]
```diff
@@ -146,4 +149,78 @@ public boolean hasChildren(Account item) {
...
         addCard("TreeGrid with lazy loading", grid);
     }
+
+    // Drag and drop
+    private Department draggedItem;
+
+    private void createDragAndDrop() {
+        DepartmentData departmentData = new DepartmentData();
+        // begin-source-example
+        // source-example-heading: Drag and drop
```
Comment (tmattsso) [inline vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java]: There's no comments on any of this code. At minimum document what part of the code does what.

File: vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java [hunk 2]
```diff
@@ -146,4 +149,78 @@ public boolean hasChildren(Account item) {
...
+        TreeGrid<Department> treeGrid = new TreeGrid<>();
+
+        treeGrid.setDataProvider(
+                new AbstractHierarchicalDataProvider<Department, Object>() {
```
Comment (tmattsso) [inline vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java]: Is this code necessary for this demo? Consider abstracting it out to a class or method.

File: vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java [hunk 3]
```diff
@@ -146,4 +149,78 @@ public boolean hasChildren(Account item) {
...
+        treeGrid.addHierarchyColumn(Department::getName)
+                .setHeader("Department Name");
+        treeGrid.addColumn(Department::getManager).setHeader("Manager");
+        treeGrid.setRowsDraggable(true);
+        treeGrid.setDropMode(GridDropMode.ON_TOP);
+
+        treeGrid.addDragStartListener(
+                event -> draggedItem = event.getDraggedItems().get(0));
+
+        treeGrid.addDragEndListener(event -> draggedItem = null);
```
Comment (tmattsso) [inline vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java]: draggedItem should be part of the code shown to the user

File: vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java [hunk 4]
```diff
@@ -146,4 +149,78 @@ public boolean hasChildren(Account item) {
...
+
+        treeGrid.addDropListener(event -> {
+            Department department = event.getDropTargetItem().get();
```
Comment (tmattsso) [inline vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java]: the variable needs a better name, it wasn't immediately clear what this code was doing.

Is there a way of preventing this drop action completely?  with specifying allowed drop targets or something?

File: vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/data/DepartmentData.java [hunk 5]
```diff
@@ -8,7 +8,7 @@
 import java.util.stream.Collectors;
 
 public class DepartmentData {
-    private static final List<Department> DEPARTMENT_LIST = createDepartmentList();
+    private final List<Department> DEPARTMENT_LIST = createDepartmentList();
```
Comment (tmattsso) [inline vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/data/DepartmentData.java]: Why remove 'static'?

Comment (shadikhani) [reply vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/data/DepartmentData.java]: Previously, the grid was read-only, but now it is not. The parent of the dragged items is changing by dragging and dropping items.

File: vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java (hunk 1 above)
Comment (shadikhani) [reply vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java]: Done.

File: vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java (hunk 2 above)
Comment (shadikhani) [reply vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java]: Done.

File: vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java (hunk 3 above)
Comment (shadikhani) [reply vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java]: Done.

File: vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java (hunk 4 above)
Comment (shadikhani) [reply vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java]: Variable name changed.

I have no idea.

**Comment by tomivirkki** [comment]
This PR's base is `master`, which is the V15 version of the component. vcom runs V14 only so the changes of this PR wouldn't end up on the website. We're planning to remove the demo-modules from the `master` branches on some timeframe as they serve no purpose currently. You can change base with the "Edit" button next to the title.

**Comment by tomivirkki** [comment]
The new demo feels buggy.

Kapture 2020-03-31 at 12 38 49

Also, there are D&D demos with TreeGrid already at:

https://github.com/vaadin/vaadin-grid-flow/blob/master/vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/grid/demo/GridDemo.java#L2364-L2428

https://github.com/vaadin/vaadin-grid-flow/blob/master/vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/grid/demo/GridDemo.java#L2526-L2605

Those could be reused (or copied) in TreeGridDemo.

But still, there are lots of Grid features that aren't demonstrated in the TreeGrid demos (like (multi)selection, sorting, filtering, header/footer config, component renderer, item details, context menu, editor...). Why should D&D be an exception? Wouldn't it make sense that the TreeGrid demos only showcase features specific to TreeGrid? Maybe TreeGrid demos could have a link to Grid demos to emphasize that all Grid features also work with TreeGrid.

**Comment by tomivirkki** [comment]
Closing as inactive.

**Pull request closed without merging**